R/variable-derivation.R

```diff
@@ -0,0 +1,82 @@
+#' Get or set a derived variable's `CrunchExpr`
+#'
+#' Get a derived variable's derivation formula as a (`CrunchExpr`)[expressions] with
+#' `derivation([variable])`. Set (change) a derived variable's derivation with
+#' `derivation([variable]) <- [expression]`
+#'
+#' @param x a variable
+#' @param value a `CrunchExpr` to be used as the derivation (for the setter only)
+#'
+#' @return a `CrunchExpr` of the derivation
+#'
+#' @examples
+#' \dontrun{
+#'
+#' ds$derived_v1 <- ds$v1 + 5
+#'
+#' derivation(ds$derived_v1)
+#' # Crunch expression: v1 + 5
+#'
+#' derivation(ds$derived_v1) <- ds$v1 + 10
+#' derivation(ds$derived_v1)
+#' # Crunch expression: v1 + 10
+#' }
+#' @name derivations
+#' @aliases derivation derivation<-
+NULL
+
+#' @export
+#' @rdname derivations
+setMethod("derivation", "CrunchVariable", function(x) {
+    if (!is.derived(x)) {
+        return(NULL)
+    }
+    expr <- CrunchExpr(expression=entity(x)@body$derivation)
+    expr@expression <- absolutifyVariables(expr@expression, x)
+
+    return(expr)
+})
+
+absolutifyVariables <- function (expr, var) {
+    if ("function" %in% names(expr)) {
+        expr[["args"]] <- absolutifyExpressionArgs(expr[["args"]], var)
```

Comment:
Why do you need two functions to recurse through this list? My intuition is something like

```
absolutifyVariables <- function (expr, base.url) {
    if (is.list(expr)) {
        # Recurse.
        lists <- vapply(expr, is.list, logical(1))
        expr[lists] <- lapply(expr[lists], absolutifyVariables, base.url)
        # For others, look for "variable" references
        if (length(names(expr))) {
            vars <- names(expr) %in% "variable"
            expr[!lists & vars] <- lapply(expr[!lists & vars], absoluteURL, base.url)
            return(expr)
        }
    } else {
        return(expr)
    }
}
```

Does that work? I haven't run it, just typed through the logic.